subsys/net/lib/nrf_cloud/src/nrf_cloud_jwt.c

```diff
@@ -8,12 +8,451 @@
+				   __FILE__, sizeof(__FILE__));
+
+	if (res != TC_CRYPTO_SUCCESS) {
+		LOG_ERR("tc_ctr_prng_init() failed, error: %d", res);
+		return -EINVAL;
+	}
+
+	prng_init = true;
+
+	return 0;
+}
+
+int default_CSPRNG(uint8_t *dest, unsigned int size)
+{
+	return tc_ctr_prng_generate(&prng_state, NULL, 0, dest, size);
+}
+
+static void base64_url_format(char *const base64_string)
+{
+	if (base64_string == NULL) {
+		return;
+	}
+
+	char *found = NULL;
+
+	/* replace '+' with "-" */
+	for (found = base64_string; (found = strchr(found, '+'));) {
+		*found = '-';
+	}
+
+	/* replace '/' with "_" */
+	for (found = base64_string; (found = strchr(found, '/'));) {
+		*found = '_';
+	}
+
+	/* remove padding '=' */
+	found = strchr(base64_string, '=');
+	if (found) {
+		*found = '\0';
+	}
+}
+
+static int sign_message_der(const char *const der_key, const char *const msg_in,
+			    char *const sig_out)
+{
+	struct tc_sha256_state_struct ctx;
+	uint8_t hash[ECDSA_SHA_256_HASH_SZ];
+	int res;
+
+	tc_sha256_init(&ctx);
+	tc_sha256_update(&ctx, msg_in, strlen(msg_in));
+	tc_sha256_final(hash, &ctx);
+
+	res = setup_prng();
+	if (res != 0) {
+		LOG_ERR("uECC_sign() failed, error: %d", res);
+		return res;
+	}
+
+	LOG_HEXDUMP_DBG(hash, sizeof(hash), "SHA256 hash: ");
+
+	uECC_set_rng(&default_CSPRNG);
+
+	/* Note that tinycrypt only supports P-256. */
+	res = uECC_sign(der_key, hash, sizeof(hash), sig_out, &curve_secp256r1);
+	if (res != TC_CRYPTO_SUCCESS) {
+		LOG_ERR("uECC_sign() failed, error: %d", res);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void remove_newlines(char *const str)
+{
+	size_t new;
+	size_t len = strlen(str);
+
+	for (size_t old = new = 0; old < len; ++old) {
+		if (str[old] != '\n') {
+			str[new++] = str[old];
+		}
+	}
+
+	str[new] = '\0';
+}
+
+#define BEGIN_PRV_KEY "-----BEGIN PRIVATE KEY-----"
+#define END_PRV_KEY "-----END PRIVATE KEY-----"
+static int strip_non_key_data(char *const str)
+{
+	char *start;
+	char *end;
+	size_t new_len;
+
+	/* Find and remove end string */
+	end = strstr(str, END_PRV_KEY);
+	if (end == NULL) {
+		return -EINVAL;
+	}
+	*end = '\0';
+
+	/* Find and remove begin string */
+	start = strstr(str, BEGIN_PRV_KEY);
+	if (start == NULL) {
+		return -EINVAL;
+	}
+
+	start += strlen(BEGIN_PRV_KEY);
+	new_len = strlen(start);
+
+	/* Move key data to the front */
+	memmove(str, start, new_len);
+	str[new_len] = '\0';
+
+	remove_newlines(str);
+
+	return 0;
+}
+
+static int get_key_from_cred(const int sec_tag, char *const der_out)
+{
+	static char pem[PRV_KEY_PEM_SZ];
+	static char der[PRV_KEY_DER_SZ];
+	size_t pem_sz = sizeof(pem);
+	size_t der_sz;
+	int err;
+
+	/* Get the private key from protected storage */
+	err = tls_credential_get(sec_tag, TLS_CREDENTIAL_PRIVATE_KEY,
+				 pem, &pem_sz);
+	if (err) {
+		LOG_ERR("tls_credential_get() failed, error: %d", err);
+		return err;
+	}
+
+	err = strip_non_key_data(pem);
+	if (err) {
+		LOG_ERR("Failed to parse PEM file, error: %d", err);
+		return err;
+	}
```

Comment:
Technically, the keys stored here may be either in PEM format or in DER format. MbedTLS figures it out automatically based on whether the key is NULL-terminated. You can either do this too, or let MbedTLS handle it for you. Third option is to ignore it for now, but we'll be locking ourselves out of using DER credentials, and this will potentially create a confusing error for anyone who tries to.